### PR TITLE
Add tracing logic in pallet macro for hooks and dispatchables

### PR DESCRIPTION
**Pull request by thiolliere**
Fix https://github.com/paritytech/substrate/issues/8303

decl_module add some tracing logic. This tracing is now also expanded in pallet macro.

Note that the tracing slightly differ between dispatchables from decl_module as the logic of the dispatchable function is not changed in the pallet macro.
Instead the trace is entered in the `UnfilteredDispatch` implementation (which is the implementation called when dispatching an extrinsic)

### CODE REVIEW

**Comment by emostov** [comment]
Does it make sense to add some tests to show how the traces in the new macro compare with the old macro?

**Comment by dvdplm** [comment]
cc @insipx

**Comment by bkchr** [comment]
> Test would be nice, but since I think we can and should drop support for `decl_` in the next release, fine without it as well.

If someone is relying on this stuff, we need a test for this.

**Comment by bkchr** [comment]
> like by initializing tracing_subscriber in some way which writes to some local vec instead of stderr ?

Yeah that also works.

But the issue is okay for now.

**Comment by emostov** [comment]
Just wanted to document here a difference with the spans from old vs new macro I noticed:

```
2021-03-11 11:30:24.833  TRACE tokio-runtime-worker sc_tracing: pallet_timestamp::pallet: on_initialize, time: 7194, id: 38281146588463105, parent_id: None    
2021-03-11 11:30:24.833  TRACE tokio-runtime-worker sc_tracing: pallet_babe: on_initialize, time: 43113, id: 40532946402148353, parent_id: None
```    

new macro targets: `pallet_{{pallet name}}::pallet`
old macro targets: `pallet_{{pallet name}`

Guillaume mentioned offline using something `<T::PalletInfo as frame_support::traits::PalletInfo>::name::<#pallet_ident<#type_use_gen>>()` to see if that gets the correct target name. I am going to give that a go.

**Comment by thiolliere** [comment]
if need the pallet name better to trace it explicitly instead of just tracing: `on_initialize`.
the new macro define the function in the module `pallet` that's why it is written in the tracing.

File: frame/support/src/lib.rs [hunk 1]
```diff
@@ -1229,6 +1229,8 @@ pub mod pallet_prelude {
 /// NOTE: OnRuntimeUpgrade is implemented with `Hooks::on_runtime_upgrade` and some additional
 /// logic. E.g. logic to write pallet version into storage.
 ///
+/// NOTE: The macro also add some tracing logic when implementing above traits.
```
Comment (shawntabrizi) [inline frame/support/src/lib.rs:1232]:
```suggestion
/// NOTE: The macro also adds some tracing logic when implementing above traits.
```

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
I am out of the loop on who is relying on these traces and why we need tests?

I mean tests are great, but are people taking a dependency on trace information to trigger some workflows?

**Comment by emostov** [comment]
> I am out of the loop on who is relying on these traces and why we need tests?

@shawntabrizi As of now, the current designs for balance reconciliation (and general state change auditing) rely on tracking DB read/writes via traces. A fragile approach, but it is the current approach.

Also to follow up on my earlier comments, I am happy with this PR as is. Tests are nice but given the current status quo I don't think they should be a blocker. 
As discussed offline with @dvdplm, @insipix - there may be merit in adding the instantiated pallets name (i.e. how it is named in the runtime) to the spans as values in order to correctly attribute events; however that is beyond the scope of this PR.

File: frame/support/src/lib.rs [hunk 2]
```diff
@@ -1229,6 +1229,8 @@ pub mod pallet_prelude {
 /// NOTE: OnRuntimeUpgrade is implemented with `Hooks::on_runtime_upgrade` and some additional
 /// logic. E.g. logic to write pallet version into storage.
 ///
+/// NOTE: The macro also adds some tracing logic when implementing the above traits.
+///
```
Comment (dvdplm) [inline frame/support/src/lib.rs:1233]:
```suggestion
/// NOTE: The macro also adds some tracing logic when implementing the above traits. The following
///  hooks emit traces: `on_initialize`, `on_finalize` and `on_runtime_upgrade`.
///
```

**Review by dvdplm — APPROVED** [review]
lgtm

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**